src/github_runner_image_builder/openstack_builder.py

```diff
@@ -521,3 +514,59 @@ def _wait_for_snapshot_complete(
     if not image.status == "active":
         logger.error("Timed out waiting for snapshot to be active, %s.", image.name)
         raise TimeoutError(f"Timed out waiting for snapshot to be active, {image.id}.")
+
+
+@dataclasses.dataclass
+class _UploadCloudConfig:
+    """The upload clouds arguments wrapper.
+
+    Attributes:
+        arch: The architecture of the image to use as build base.
+        image_name: The name to upload the image as.
+        keep_revisions: Number of revisions to keep before deletion.
+    """
+
+    arch: Arch
+    image_name: str
+    keep_revisions: int
+
+
+def _upload_to_clouds(
+    conn: openstack.connection.Connection,
+    image: openstack.compute.v2.image.Image,
+    upload_cloud_names: typing.Iterable[str] | None,
+    upload_cloud_config: _UploadCloudConfig,
+) -> tuple[openstack.compute.v2.image.Image, ...]:
+    """Upload the snapshot image to different clouds.
+
+    Args:
+        conn: The OpenStack connection instance.
+        image: The snapshot image to upload.
+        upload_cloud_names: The clouds to upload the image to.
+        upload_cloud_config: The upload image configuration.
+
+    Returns:
+        The uploaded cloud images.
+    """
+    if not upload_cloud_names:
+        return (image,)
```

Comment:
Perhaps just don't call the function in this case?

Reply:
This is to reduce the if/else statement on the controller function which helps reduce complexity when reading it :)

Reply:
Can you think of any alternative ways?

Reply:
What would your suggestion be? The caller function acts as a controller and delegates sub-level checks and returns values with consistent types. I think it is fine.